use-cases/NIC-passthrough-and-Kata.md

```diff
@@ -0,0 +1,292 @@
+
+
+```bash
+dmesg | grep -ie iommu
+```
+
+> Expected results
+>
+> ...
+>
+> [    1.081716] iommu: Adding device 0000:08:00.0 to group 33 
+>
+> [    1.081736] iommu: Adding device 0000:09:00.0 to group 34 
+>
+> [    1.081755] iommu: Adding device 0000:0a:00.0 to group 35 
+>
+> [    1.081776] iommu: Adding device 0000:0b:00.0 to group 36 
+>
+> [    1.081795] iommu: Adding device 0000:0c:00.0 to group 37 
+>
+> ...
+
+
+
+#### 3. Enable the *hotplug_vfio_on_root_bus* configuration in the *kata configuration.toml* file:
+
+```bash
+vi /usr/share/defaults/kata-containers/configuration.toml
+```
+a. Set hotplug_vfio_on_root_bus = true
+
+b. Make sure you are using the PC machine type by verifying machine_type = "pc"
+
+
+
+#### 4. Install *lshw* and *pciutils* (Ubuntu)
+
+```bash
+apt install lshw pciutils
+```
+
+
+
+#### 5. Find PCI address for the interface you wish to pass to the container
+
+Say I want to pass **enp0s20f1** to the container (see what interfaces you have by running `# ip link` ). 
+
+Get address: 
+
+```bash
+lshw -class network -businfo -numeric | grep enp0s20f1
+```
+
+> Expected results 
+> (something like)
+> 
+> pci@0000:00:14.1 enp0s20f1 network Ethernet Connection I354 1.0 GbE Backplane
+
+
+
+#### 6. Insert the address into a variable
+
+```bash
+BDF="0000:00:14.1"
+```
+
+
+
+#### 7. Load *vfio-pci*  kernel module /TODO maybe set this to permanent? 
+
+```bash
+modprobe -i vfio-pci
+```
+Test that the module loaded correctly: 
+
+```bash
+lsmod | grep vfio
+```
+
+> Expected results
+>
+> vfio_pci                      41268     0
+>
+> vfio_iommu_type1  22300      0
+>
+> vfio                            32656      2     *vfio_iommu_type1*,vfio_pci
+>
+> irqbypass                 13503      2      kvm,vfio_pci
+
+
+
+#### 8. Unbind the Interface from its current driver. 
+
+```bash
+echo $BDF | sudo tee /sys/bus/pci/devices/$BDF/driver/unbind
+```
+
+> Expected results
+>
+> 0000:00:14.1
+
+
+
+#### 9.     Find vendor & device ID
+
+```bash
+lspci -n -s $BDF
+```
+
+> Expected results (something like)
+>
+> 00:14.1 0200: 8086:1f40 (rev 03)
+
+
+
+#### 10.  Bind the device to vfio-pci using the result of the previous step.
+
+```bash
+echo 8086 1f40 | sudo tee /sys/bus/pci/drivers/vfio-pci/new_id
+```
+
+**Note:** the ` xxxx xxxx ` format is of the correct form and not a mistake.
+
+> Expected results (something like)
+> 
+> 8086 1f40
+
+#### 11. Check what the new device name is, now under vfio
+
+```bash
+ls /dev/vfio
+```
+
+> Expected results 
+> (something like)
+>
+> 39  vfio
+
+**Note:** that the requested outcome here is the newly created “39” 
+
+#### All set here. Time to move to the Container’s side of the deal.
+
+
+### Part 2 – The Container
+
+#### 12.   Create the container you wish to pass the NIC into
+
+**Note:** once running the next command you will be automatically transferred into the container’s bash shell. 
```

Comment:
We probably don't need this note - hopefully folks running Kata know how to run docker, and we can't explain all details to them ;-).